tests/commands/test_build.py

```diff
@@ -0,0 +1,692 @@
+    validator.basedir = tmp_path
+    resp = validator.validate_requirement(None)
+    assert resp == []
+
+
+def test_validator_requirement_default_missing(tmp_path):
+    """'requirement' param: default value when no requirements.txt is there."""
+    validator = Validator()
+    validator.basedir = tmp_path
+    resp = validator.validate_requirement(None)
+    assert resp == []
+
+
+def test_validator_requirement_absolutized(tmp_path, monkeypatch):
+    """'requirement' param: check it's made absolute."""
+    # change dir to the temp one, where we will have the reqs file
+    testfile = tmp_path / 'reqs.txt'
+    testfile.touch()
+    monkeypatch.chdir(tmp_path)
+
+    validator = Validator()
+    resp = validator.validate_requirement(['reqs.txt'])
+    testfile = pathlib.Path(str(testfile))  # comparison below don't work well in Py3.5
+    assert resp == [testfile]
+
+
+def test_validator_requirement_expanded():
+    """'requirement' param: expands the user-home prefix."""
+    home = pathlib.Path("~").expanduser()
+    testfd, testpath = tempfile.mkstemp(dir=str(home))
+    try:
+        testpath = pathlib.Path(testpath)
+
+        validator = Validator()
+        relative_testpath = "~/{}".format(testpath.relative_to(home))
+        resp = validator.validate_requirement([relative_testpath])
+    finally:
+        testpath.unlink()
+    assert resp == [testpath]
+
+
+def test_validator_requirement_exist():
+    """'requirement' param: checks that the file exists."""
+    validator = Validator()
+    expected_msg = "the requirements file was not found: '/not_really_there.txt'"
+    with pytest.raises(CommandError, match=expected_msg):
+        validator.validate_requirement(['/not_really_there.txt'])
+
+
+# --- Polite Executor tests
+
+def test_politeexec_base(caplog):
+    """Basic execution."""
+    caplog.set_level(logging.ERROR, logger="charmcraft")
+
+    cmd = ['echo', 'HELO']
+    retcode = polite_exec(cmd)
+    assert retcode == 0
+    assert not any("Execution ended" in rec.message for rec in caplog.records)
+
+
+def test_politeexec_stdout_logged(caplog):
+    """The standard output is logged in debug."""
+    caplog.set_level(logging.DEBUG, logger="charmcraft")
+
+    cmd = ['echo', 'HELO']
+    polite_exec(cmd)
+    assert [":: HELO"] == [rec.message for rec in caplog.records]
+
+
+def test_politeexec_stderr_logged(caplog):
+    """The standard error is logged in debug."""
+    caplog.set_level(logging.DEBUG, logger="charmcraft")
+
+    cmd = ['python3', '-c', "import sys; print('weird, huh?', file=sys.stderr)"]
+    polite_exec(cmd)
+    assert [":: weird, huh?"] == [rec.message for rec in caplog.records]
+
+
+def test_politeexec_failed(caplog):
+    """It's logged in error if cmd fails."""
+    caplog.set_level(logging.ERROR, logger="charmcraft")
+
+    cmd = ['python3', '-c', "exit(3)"]
+    retcode = polite_exec(cmd)
+    assert retcode == 3
+    assert any("Execution ended in 3" in rec.message for rec in caplog.records)
```

Comment:
"Execution ended in 3" doesn't feel very useful to give as information to users. Maybe its 

"Executing {cmd} failed with return code {retcode}" 
?

Reply:
Nice! I also improved the message when crashing...